spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java

```diff
@@ -0,0 +1,306 @@
+		ModuleDeploymentId id = ModuleDeploymentId.fromModuleDefinition(request.getDefinition()); 
+		
+		logger.debug("Deploying module: {}", createKubernetesName(id));
+		
+		Integer externalPort = 8080;
+		// we also create a service in case have an '--port' or '--server.port' argument (source or sink)
+		Map<String, String> parameters = request.getDefinition().getParameters();
+		if (parameters.containsKey(PORT_KEY)) {
+			externalPort = Integer.valueOf(parameters.get(PORT_KEY));
+			createService(id, request, externalPort);
+		} else if (parameters.containsKey(SERVER_PORT_KEY)) {
+			externalPort = Integer.valueOf(parameters.get(SERVER_PORT_KEY));		
+			createService(id, request, externalPort);
+		}		
+		createReplicationController(id, request, externalPort);				
+		return id;	
+	}
+
+	@Override
+	public void undeploy(ModuleDeploymentId id) {
+		String name = createKubernetesName(id);
+		logger.debug("Undeploying module: {}", name);
+
+		try {
+			kubeClient.replicationControllers().withName(name).delete();		
+			kubeClient.services().withName(name).delete();
+		} catch (KubernetesClientException e) {
+			logger.error(e.getMessage(), e);
+			// FIXME what do we throw??
+		}
+	}
+
+	@Override
+	public ModuleStatus status(ModuleDeploymentId id) {
+		String name = createKubernetesName(id);
+		logger.debug("Querying module status: {}", name);
+
+		try {
+			// The only really interesting status is coming from the containers and pods.
+			// The service and the RC don't have "realtime" status info.
+			PodList list = kubeClient.pods().withLabels(createIdMap(id)).list();
+			return buildModuleStatus(id, list);
+		}
+		catch (KubernetesClientException e) {
+			logger.warn(e.getMessage(), e);
+			return buildModuleStatus(id, null);
+		}
+	}
+	
+	@Override
+	public Map<ModuleDeploymentId, ModuleStatus> status() {
+		Map<ModuleDeploymentId, ModuleStatus> result = new HashMap<>();
+		
+		ReplicationControllerList list = 
+				kubeClient.replicationControllers()
+				.withLabel(SPRING_MARKER_KEY, SPRING_MARKER_VALUE).list();
+		
+		for (ReplicationController rc : list.getItems()) {
+			Map<String, String> labels = rc.getMetadata().getLabels();			
+			String group = labels.get(SCSM_GROUP_KEY);
+			String label = labels.get(SCSM_LABEL_KEY);			
+			ModuleDeploymentId id = new ModuleDeploymentId(group, label);
+			
+			PodList pods = kubeClient.pods().withLabels(labels).list();			
+			result.put(id, buildModuleStatus(id, pods));	
+		}		
+		return result;		
+	}
+
+	private ReplicationController createReplicationController(ModuleDeploymentId id, ModuleDeploymentRequest request, Integer externalPort) {
+        ReplicationController rc = new ReplicationControllerBuilder()
+                .withNewMetadata()
+                	.withName(createKubernetesName(id))  // does not allow . in the name
+                	.withLabels(createIdMap(id))
+                	.addToLabels(SPRING_MARKER_KEY, SPRING_MARKER_VALUE)
```

Comment:
Are all these really necessary when it seems the ones returned from `createIdMap()` are the only ones used to `filter` s-c-d modules?